doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1323 @@
+
+Each section will begin with a short motivation of the issue or what
+problem is addressed. This is followed by a description of the proposed
+design choice, and then may list alternatives.
+
+
+Nomenclature
+""""""""""""
+
+As a brief note on nomenclature, it should be noted that ``dtype`` normally
+denotes the dtype *instance*, which is the object attached to a numpy array.
+On the other hand the ``DType`` class is the subclass of ``np.dtype``.
+On the C-level we currently use the name ``descriptor`` or ``descr`` in many
+cases interchangeably with the dtype instance. This will be used in proposed
+C-API names for clarification.
+Note that the notion of dtype class is currently represented mainly as
+the ``dtype.num`` and ``dtype.char``.
+Please see the `dtype hierarchy figure <hierarchy_figure>`_ for an
+illustration of this distinction.
+
+There are currently classes in NumPy for numeric types e.g. 
+``np.float64``; however,
+these are not DTypes but the corresponding scalar classes
+(see also NEP 40 and 41 for discussion on why these are largely unrelated to
+the proposed changes).
+
+
+Proposed access to DType class
+""""""""""""""""""""""""""""""
+
+**Motivation:**
+
+Currently we often use e.g. ``np.dtype(np.int64)`` to create the dtype
+instance corresponding to a given scalar type. Adding the DType class
+means that a way to access that class may be required.
+
+**Description:**
+
+To avoid duplication, but also expose the classes conveniently to users
+we propose the addition of::
+
+    np.dtype[np.int64]
+
+as a class getter. This can work both for user and NumPy DTypes,
+although, in many cases a library may choose to provide a more direct
+way to access the specific DType class.
+This method may initially be limited to concrete DTypes. 
+The main reason for this choice is to provide a single
+clear and future-proof way to find the DType class given the
+Python (scalar) class.
+
+This should not be a common operation, so providing this helper reduces the
+pressure of adding the new DType classes into the namespace.
+
+
+Hierarchy of DTypes and Abstract DTypes
+"""""""""""""""""""""""""""""""""""""""
+
+**Motivation:**
+The creation of a DType classes has already been decided in NEP 41.
+Here we discuss the notion of **abstract** DTypes.
+There are multiple reasons for this:
+
+1. It allows the definition of a class hierarchy, in principle allowing checks like
+   ``isinstance(np.dtype("float64"), np.inexact)``.
+   **This hierarchy may be a prerequisite to implementing dispatching
+   for universal functions (NEP 43)**
+2. Abstract DTypes can enable code such as
+   ``arr.astype(Complex)`` to express the desire to cast to a
+   complex data type of any precision.
+3. It anticipates the creation of families of DTypes by users.
+   For example allowing the creation of an abstract ``Unit`` class with a concrete
+   ``Float64Unit``. In which case ``Unit(np.float64, "m")`` could be
+   identical to ``Float64Unit("m")``.
```

Comment:
Well, I think its open for the implementer, and I agree that `Unit(np.float64, "m")` should give a concrete `dtype` instance (I assume this is what you mean).  I personally think the existence of a `Float64Unit` as a DType class may make sense, because it seems to me it would make dispatching simpler (and a single step).  However, I agree that this is not strictly necessary. I will try to shorten the example a bit.